JBPM-5462: Stunner - Memory leaks

romartin opened the pull request:
Hey @hasys @manstis @tiagodolphine 

Here is the PR that fixes the most important memory leaks. These are the related tickets:
- https://issues.jboss.org/browse/JBPM-5462
- https://issues.jboss.org/browse/JBPM-6330

Here are the most important changes:

**Fixed leaks on lienzo side**
The two most relevant were on the LIenzoHandlerManager and on the Pictures - some handlers and DOM elements were added but never removed again, this was causing the toolbox, and any other component that uses lienzo pictures (images), to be leaking on each rendering operation.

**Fixed leaks in Stunner "factories"**
In general removed most of the  Stunner "factory" types and implementations, like session factories, presenters factories, canvas factrories, control facorires, etc. 
This way refactored the different beans, which were produced by those factories, into injectable CDI beans, so the scope for that beans, and its resource consumption, its being managed by the dependent instance scope as well

**Fixed leaks in Stunner client components - sessions and controls** 
Renamed `ClientReadOnlySession` and `ClientFullSession` by `ViewerSession` and `EditorSession`. 
Also removed all factories and created "managed" implementations, like the `ManagedSession`, which obtains the right CDI instances for each of the session components at runtime.

**Fixed leaks in other Stunner widgets, presenters, screens and editors**
This way the editors and screens are now creating and destroying all the instances given their life-cycle methods (onOpen, onClose, etc)

- Refactored the `ClientSessionCommands` and the use in the editors..

- Changed also some scope for beans and lots of other small changes in several components to ensure they're destroying their states, and the states for the dependent instances...

**NOTICE** there is still some missing work to do on the testing side - there are a few  `@Ignored` tests, others are still missing to be refactored, missing tests for the lienzo side too. and in general are missing tests for some of the new classes in Stunner, will keep working on this tomorrow, but at least this way you can start the review, as I expect it can take a few days.

**PS #1**: I don't say all the memory leaks are fixed, but it works fine after opening,saving, close, creating and reopening several diagrams... so please take this into account  it doesn't make sense to start profiling the application and identify the missing leaks, the goals for this task are making Stunner good enough for a product release. 

**PS #2**: Next step is to improve the performance on some computations, as you'll see that initially it takes like 1-2sec in some operations to complete. Will try to work on this during next week.

**PS #3**: I'm sorry for the size of the PR! But due to removing factories and client session implementations in general the number of changed files was grown so fast... 

**Related PRs**:
- https://github.com/kiegroup/lienzo-core/pull/21
- https://github.com/kiegroup/lienzo-tests/pull/18

Thanks!

manstis commented:
@romartin This PR is already in need of conflict fixes and will probably require more after https://github.com/kiegroup/kie-wb-common/pull/1680 is merged.. I propose you revisit this PR when the former has been merged (as opposed to fixing conflicts now and have to re-do when #1680 is merged).

romartin commented:
@manstis yeah makes sense, I'll work today on missing tests and once yours merged, I'll do the rebase, np! Thx!

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/ProjectEditorMenuSessionItems.java:
```diff
@@ -0,0 +1,275 @@
+    }
+
+    private void setItemEnabled(final Class<? extends ClientSessionCommand> type,
+                                final boolean enabled) {
```
This parameter value of private method is always `False`

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/session/EditorSessionCommands.java:
```diff
@@ -0,0 +1,148 @@
+    }
+
+    public VisitGraphSessionCommand getVisitGraphSessionCommand() {
+        return commands.get(0);
```
Does it make sense to use Enumerators instead of magic numbers here?

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/screens/SessionDiagramEditorScreen.java:
```diff
@@ -301,17 +281,12 @@ public void onError(final ClientRuntimeError error) {
-            log(FINE,
-                "Session already active, no action.");
-        }
+        // TODO: presenter.resume();
```
Another new TODO, should we report a Jira issue to keep it track?

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-showcase/kie-wb-common-stunner-showcase-standalone/src/main/java/org/kie/workbench/common/stunner/standalone/client/screens/SessionDiagramEditorScreen.java:
```diff
@@ -320,29 +295,19 @@ private boolean isSameSession(final ClientSession other) {
     @OnLostFocus
     public void OnLostFocus() {
-
+        // TODO presenter.pause();
```
Another new TODO, should we report a Jira issue to keep it track?

hasys commented on kie-wb-common-stunner/kie-wb-common-stunner-client/kie-wb-common-stunner-widgets/src/main/java/org/kie/workbench/common/stunner/client/widgets/toolbar/impl/ManagedEditorToolbar.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    public VisitGraphToolbarCommand getVisitGraphToolbarCommand() {
+        return (VisitGraphToolbarCommand) toolbar.getCommand(0);
```
Again magic numbers, I guess all of them are represents the same meaning. If yes, it is better to use Enumerator for sure.

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/session/presenters/impl/DMNSessionPreview.java:
```diff
@@ -0,0 +1,106 @@
+
+@DMNEditor
+@Dependent
+public class DMNSessionPreview implements SessionDiagramPreview<AbstractSession> {
```
There is no injection point for it at the code right now, as it's being injected at runtime, but it's being used (notice the use of the annotation `@DMNEditor`)

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/session/presenters/impl/SessionPreviewImplTest.java:
```diff
@@ -16,50 +16,17 @@
 @RunWith(MockitoJUnitRunner.class)
 public class SessionPreviewImplTest extends BaseCommandsTest {
-
+/*
```
Fixed on my next commit

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-project-client/src/main/java/org/kie/workbench/common/dmn/project/client/editor/DMNDiagramEditor.java:
```diff
@@ -71,6 +66,14 @@
 
     private final DecisionNavigatorDock decisionNavigatorDock;
 
+    /*
```
Fixed

romartin commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/widgets/toolbar/DMNEditorToolbar.java:
```diff
@@ -29,38 +31,98 @@
+    }
+
+    public VisitGraphToolbarCommand getVisitGraphToolbarCommand() {
+        return (VisitGraphToolbarCommand) toolbar.getCommand(0);
```
yes good point, let's keep it as improvement plz... for now, as you can use the getters for each command, the magic number in fact keeps "hidden" from the API, so it's not so bad.... :) but yeah agree, will add to my list ;)

romartin commented:
Fixed comments from @hasys , fixed the issue that was causing not being able to open multiple editors, and rebased with master.

STILL WORK IN PROGRESS

romartin commented:
Jenkins please retest this

romartin commented:
Hey @manstis @hasys @tiagodolphine 

I've been working on fixing the blocker issues on this PR.  In general I've been testing and everything seems fine, this commit is really important please, I've rebased with master it's ready for testing, so would appreciate if you can give it priority and we can try to merge it ASAP.

This is the status:
- Fixed some more tests
- Fixed the keyboard events (were not working before)
- Fixed multi-session - now it's possible to open multiple processes

Tomorrow at the Stunner's meeting I'll give some updates on this, there are good news and bad news. Good news are about the leaks are almost completely being fixed, the bad news is that working with more than 4 processes at same time is making the tool slow down... but that's not an Stunner leak, will give you more details at the call, so we can go ahead with this work anyway and try to merge it.

Also notice there are missing a few tests... will keep working on these during next days.

Thanks!
Roger

romartin commented:
@hasys I've also merged your PR and fixed your comments. Thx!

romartin commented:
Jenkins please retest this

romartin commented:
**LAST UPDATE** 
it seems not so much "bad news" - it's probably a chrome bug! Stunner in FF does not slows down when having multiple editors open... it only happens in chrome

manstis commented:
Jenkins please retest this.

manstis commented:
Jenkins please execute full downstream build.

tiagodolphine commented:
@romartin I noticed one issue, if you open one diagram insert some nodes and than open a new diagram insert some nodes and delete, copy/paste using keyboard shortcuts and after this select the previous diagram and try to delete any node using the keyboard,  you will not be able, the shortcuts doesn't work on both opened diagrams.

romartin commented:
@tiagodolphine  mmmm good point, lemme give it a try

romartin commented:
Hey @hasys @manstis @tiagodolphine 

I've rebased this branch again. Also I've fixed the issue found by @tiagodolphine about keyboard events on multiple editors (good catch!). I've added a few more tests but still need to work on some other commented ones (due to some refactor), but would appreciate if we can move this forward meanwhile as it's a blockers and needs to be cherry-picked ASAP. I've now tested again and keyboards and sessions are behaving fine. 

Please appreciate if you can you give priority to this, I'll focus on working on your comments, and try to complete the review soon so we can merge it...

Thanks in advance!
Roger

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin,

Sorry, but I need little bit more time there. I got some Null Pointers during playing with build, but not found a stable reproducer for it. Also, will try to play with master branch as well to be sure that those Nullprointers are related to this PR.

romartin commented:
Hey guys,

I'm still doing tests here and trying to fix an issue I see on the meanwhile you're doing your review. 

Some updates:
* Fixed a NPE that was showing up once the toolbox was disappearing, but on the meanwhile you was trying to focus it again
* Removed the max width/height constraints limitations for lanes and subprocesses (I promised you @hasys  to do this in the next commit ;) )

Will continue doing more tests on the UI for now to ensure this commit does not brokes any other stuff. Please let's try to move this all together ASAP.

Thx!

romartin commented:
hey guys,

Found another bug that was happening only when the Tree Explorer is open, it was causing an exception (sometimes) when adding or removing children from/to a parent. Added a new commit and test for it.
PS: @hasys probably those two bug fixes just added are one of the NPE you mention? Otherwise lemme know, I'm still doing more tests on the UI too..

Thanks!

romartin commented:
Jenkins execute full downstream build

romartin commented:
Hey guys,

Just rebased again (so force pushed) due to a compilation issue... everything is fine again.

Jenkins execute full downstream build

domhanak commented:
@romartin We had a test day using a snapshot with this PR and I encountered following possible issue. Here are my steps to reproduce it, could you please take a look?
1. Open a process in one Stunner editor - Model: Start -> Script -> End and Open "Diagram properties"
2. Do not close anything but navigate back to your project and create a new process.
3. In this new process model something and open "Diagram properties"
4. An error is displayed and diagram properties from previous editor is shown. Clicking the properties editor produces same exception over and over again.

NOTE: I was unable to reproduce it with ER4.

hasys commented:
Hi @romartin, than you, yes, your description of issues are very similar to what I saw. I tried now and didn't met any NullPointers any more. It looks like everything is fine now. But I found one more interesting issue:
* Open 3 Stunners
* Third one will be extremely slow, but two first will be OK
* Close 1 and 2, third process will be still extremely slow
* Open one more Process (4-th), it will be OK
* Open one more (5-th) - this will be extremely slow as well as 3-rd, but 4-th one will be OK

Not sure, should we fix it in this PR or merge it and report separate issue. I guess, it will be better to merge and fix found issue as memory leaks part 2 (as we did it for the palette).

Also, now will take a look on the code now.

hasys commented:
Hi @romartin,
revised the code, looks good to me, but I found some TODOs and disabled tests (mentioned above but some disabled tests arrived after review as well). I guess after tests and TODOs fixes (move TODOs from code to Jira is also an option, but not for a tests) this PR is ready to go. Great job, Roger!

romartin commented:
Hey @domhanak @hasys 

Thanks for the review, will work now on the issue you just found. 

About what @hasys says about performance, yeah I noticed too! It's really weird, but this is a chrome specific issue, if you do same steps in FF or Safari everything is fine.. We still have no idea why this happens on Chrome, we're still researching if ti's a chrome bug or we can do some fix, but for now the idea is to merge this, leaks are gone, and then we'll see how to address the chrome specific issue, there are re a couple of solutions...let's comment them in the next Stunner meeting and see how we solve it.

Rather than this, will work on the bug you found and some missing TODOS and will update the PR. thanks guys!

romartin commented:
Hey @domhanak @hasys 

I've been working on your comments, here is the status:

* @domhanak  I've been not able to reproduce your issue using the kie-wb :( I've tried by closing, not closing, re-opening, and nothing produces an NPE for me. Not sure if you can reproduce it with latest version. @hasys can you?  Anyway if it appears, what about reporting it and I'll address next week? Well lemme know, it depends on how blocker it could be.
* Probably makes sense to report the issue on chrome that you mention @hasys , anyway let's discuss what to do on next meeting yeah
* I've managed to refactor the missing tests for the editors and make them pass
* I know there are some missing tests mostly on the lienzo side, please appreciate if I can add this in a future commit next week...

So just added a new commit for the tests. Please it's **REALLY IMPORTANT** trying to merge this today (Fri) into master. I would suggest that if you find some bug or improvement, or whatever missing, if not absolutely a blocker, please report it and let's try to move this forward. It will be easier to manage next week "small" commits (for any bug that could appear due to this stuff) than keeping this PR open...

**The goal (ideally) is having this merged on 7.7.x as for next Monday**, so if this can be merged today, I'll be able to just cherry-pick and verify the build during the weekend, and if everything is good as on master, I'll be able to integrate this into next 7.7.x build (Mon). Does that sounds good for you? Or does somebody offers to test the 7.7.x during the weekend?? (hehe). Anyway np, I can do that, if you agree :+1: 

Thanks!

romartin commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin, mentioned above issue reproducible only with Chrome and only on Linux (I have been tested it with IE11 on Windows 8, Firefox on Windows/Linux and tested Chrome on Linux/Windows/Mac OS X). I guess it is not an issue to solve in this PR. I reported it to separate Jira: https://issues.jboss.org/browse/RHPAM-922

But I found tree more issue.
1. For IE 11 - some operations on the canvas cause NullPointer messages (you can close dialog and proceed you work, but it is very annoying because it appears very often on basic operations).
2. When you click and hold mouse out of canvas (for example on Palettes empty space) and release mouse on canvas - you will see error message. (not critical, can be fixed with separate PR).
3. It looks like this PR works little bit slower, than latest snapshot. But as we discussed it before on weekly meeting there will be some optimizations right after this PR, so I guess it is not critical for this PR as well. But we need a Jira to track optimization changes.

So, will check test changes now and it will be ready to go if tests are fine.

romartin commented:
Ok guys, thanks for the reviews! 
Just rebased and squashed, waiting now for Jenkins to build....

romartin commented:
@hasys just created [JBPM-7235](https://issues.jboss.org/browse/JBPM-7235) in order to not forget about the tests :+1:

romartin commented:
Jenkins execute full downstream build

romartin commented:
Jenkins execute full downstream build

romartin commented:
Rebased, Jenkins happy and I've just tested the WAR, everything is fine, merging.
